### PR TITLE
Close channel if a remote close event was received

### PR DESCRIPTION
**Pull request by jruesga**
You should also consider https://github.com/jruesga/chromecast-java-api-v2/commit/aff647e67a0914eb1346905e3484aaea79c95af9. If builds against 1.7 (1.7 is required to build against jitpack) and uses jcenter as main repo, amend update some libraries.

### CODE REVIEW

**Comment by vitalidze** [comment]
Thanks for the pull request. I will be on vacation within next two weeks. So will get back to this after I return.

File: src/main/java/su/litvak/chromecast/api/v2/Channel.java [hunk 1]
```diff
@@ -424,8 +441,13 @@ public Status launch(String appId) throws IOException {
...
 
     public Status stop(String sessionId) throws IOException {
-        StandardResponse.Status status = sendStandard("urn:x-cast:com.google.cast.receiver", StandardRequest.stop(sessionId), DEFAULT_RECEIVER_ID);
-        return status == null ? null : status.status;
+        stopRequested = true;
+        try {
+            StandardResponse.Status status = sendStandard("urn:x-cast:com.google.cast.receiver", StandardRequest.stop(sessionId), DEFAULT_RECEIVER_ID);
+            return status == null ? null : status.status;
+        } finally {
+            stopRequested = false;
```
Comment (vitalidze) [inline src/main/java/su/litvak/chromecast/api/v2/Channel.java]: It looks unreliable. From my understanding the message about closing will be received asynchronously. There is no guarantee that you will reach this code after you receive the 'close' response and thus you will most probably not process the close request correctly.

File: src/main/java/su/litvak/chromecast/api/v2/Channel.java [hunk 2]
```diff
@@ -159,8 +164,20 @@ public void run() {
...
                                     }
                                 }
-                            } else if (parsed.has("responseType") && parsed.get("responseType").asText().equals("PING")) {
+                            } else if (message.getNamespace().equals("urn:x-cast:com.google.cast.tp.heartbeat")
+                                    && parsed.has("responseType") && parsed.get("responseType").asText().equals("PING")) {
                                 write("urn:x-cast:com.google.cast.tp.heartbeat", StandardMessage.pong(), DEFAULT_RECEIVER_ID);
+                            } else if (message.getNamespace().equals("urn:x-cast:com.google.cast.tp.connection")
+                                    && parsed.has("responseType") && parsed.get("responseType").asText().equals("CLOSE")) {
+                                // Determine whether the close event was requested by this sender
+                                ((ObjectNode) parsed).put("requestedBySender", stopRequested);
```
Comment (vitalidze) [inline src/main/java/su/litvak/chromecast/api/v2/Channel.java]: This looks tricky. Is there some additional info in 'Close' message about the sender? Maybe there is a sender id?

Comment (jruesga) [inline src/main/java/su/litvak/chromecast/api/v2/Channel.java]: No. IIRC, it receives this CLOSE event when the Backdrop starts. There is no senderId in that moment though.

File: src/main/java/su/litvak/chromecast/api/v2/Channel.java (hunk 1 above)
Comment (jruesga) [inline src/main/java/su/litvak/chromecast/api/v2/Channel.java]: Not sure, I able to receive the event an CLOSE the app when it happens.

**Comment by jruesga** [comment]
About the 1.7 switct, was mainly to allow the library build throught Jitpack and use as a remote dependency in gradle, without the need to have it locally installed.

**Comment by jruesga** [comment]
I have this patches in my repo, which is the one I'd linked to, so I can abandone this PR, if it doesn't matches your expectations. It helps me to deal with properly close the HTTP server when my app stop being the active.

File: src/main/java/su/litvak/chromecast/api/v2/Channel.java (hunk 2 above)
Comment (vitalidze) [inline src/main/java/su/litvak/chromecast/api/v2/Channel.java]: So how does the "Close" event body looks like?

File: src/main/java/su/litvak/chromecast/api/v2/Channel.java (hunk 1 above)
Comment (vitalidze) [inline src/main/java/su/litvak/chromecast/api/v2/Channel.java]: I think you are right and my bad that I have actually missed the fact that request is posted synchronously.

**Comment by vitalidze** [comment]
> I have this patches in my repo, which is the one I'd linked to, so I can abandone this PR, if it doesn't matches your expectations. It helps me to deal with properly close the HTTP server when my app stop being the active.

I don't get the relation. How does dealing with HTTP server is related with using of jitpack?

Anyway, I would like to keep 1.6 version of java since it is still widely used.

**Comment by jruesga** [comment]
Sorry, I mixed comments though.

The "HTTP server" comment is for the CLOSE event (i have an embedded http server in my app to serve the content to the CC). I saw this event happen when backdrop starts, and handle it allow me to safelly shutdown the the http server and the other resources associated.

The 1.7 is unrelated to PR (as comment it just allow me to build the library with gradle without have it locally installed)

**Comment by maggu2810** [comment]
About the closed event...
If I understand this PR correctly there is also an event if the sender or received initiate a close.
I added some time ago code to detect a closed connection (so, no initiated close, but perhaps a disconnect / socket loss).
Is this something that should be cooperated here?
IIRC the argument has been that this is no spontaneous event and should not be added to that listener interface.
I handle this ATM this way: https://github.com/maggu2810/chromecast-java-api-v2/commit/492d9810799ffbe9eedacd8771f77ecb2a661586

**Comment by jruesga** [comment]
Per above discussion, this probably shoudn't be in the upstream repo.

**Pull request closed without merging**